### ISSUE TITLE
Migrate from SW-Precache to Workbox

### ISSUE THREAD

**Issue by latin-panda** [medic/cht-core#6514]
**What feature do you want to improve?**
We have some service workers made with SW-Precache but this library is obsolete already. 

**Describe the improvement you'd like**
SW-Precache evolved to the now called Workbox, which has new features that could be useful in the future, additionally it's good to keep our project's technology updated for better support.

**Describe alternatives you've considered**
This link could be a good guide during the migration: https://developers.google.com/web/tools/workbox/guides/migrations/migrate-from-sw

**Additional context**
Additionally, we would like to test the Workbox feature of [preloading](https://developers.google.com/web/tools/workbox/modules/workbox-navigation-preload) while having HTTP/2 enabled, considering that some regions have slow internet connection, it would be beneficial to improve the initial load of resources even though we are [precaching](https://developers.google.com/web/tools/workbox/modules/workbox-precaching) them with the worker.
This issue came from the research about [improving the initial load of login page by using preload and HTTP/2](https://github.com/medic/cht-core/issues/6276), where it can give a background context and the experience during the research. 
There's another issue about [generating the workers during runtime instead of build time](https://github.com/medic/cht-core/issues/6338), it could potentially take place after this migration is done.

**Comment by kennsippell** [comment]
Some relevant information on why sw-precache was used and not workbox: https://github.com/medic/cht-core/pull/5219

## swPrecache vs Workbox 
It is worth checking out the [Support Details for swPrecache](https://github.com/GoogleChromeLabs/sw-precache#support). TLDR: "We would recommend Workbox for new projects" but "we’ll continue to support sw-precache". 
 
Workbox looks awesome. Most notably check out [enabling offline Google Analytics](https://developers.google.com/web/tools/workbox/modules/workbox-google-analytics). So why isn't this PR using Workbox? The short answer is we probably should – but this is a big PR and Workbox is going to complicate things.  
 
Here is why I'm using the swPrecache library: 
 
1. **The community support isn't perfect for either option** - Workbox 3 supports Chrome >51, but modern [Workbox 4 only supports Chrome >56](https://github.com/GoogleChrome/workbox/releases) so really both swPrecache and latests Workbox are [both not supporting us](https://www.youtube.com/watch?v=bVwh8PW6r88). 
1. **Partial Precache State** - SwPrecache is pretty much exactly what we want by default: it fails to install the service worker if any resource doesn't get precached and thus the possibility of a partial precached state. Workbox delivers a best effort precache and it isn't immediately obvious to me how to best avoid a partial precache state. 
1. **Dependent Scripts** - SwPrecache inlines all dependencies directly in the service worker so we just serve one script and we are done. Workbox has a modular build which deploys dependencies along-side your service worker (~20 files). A bit messy and not a perfect fit for our build script which ends up adding twenty more attachments to the database. A quick attempt to build only the required resources ended up with some dependencies being pulled from a cdn – which is problematic for projects using domain-based data contracts.  
 
All solvable problems, but a bit of untangling to do. Imo -- the best case would be using Workbox to output a script which does exactly what swPrecache is outputting now. I'm confident we can get there if we are interested in other features of Workbox *cough* - offline google analytics - *cough*.

**Comment by garethbowen** [comment]
Based on the supported browsers for Workbox 4 we might need to wait for a replacement for Crosswalk before we can make the switch.

**Comment by garethbowen** [comment]
Now that we have updated our browser support policies and dropped crosswalk we should investigate this again. Points 2 and 3 above will still need to be resolved.

**Comment by garethbowen** [comment]
### Test plan

1. Log in to https://gamma.dev.medicmobile.org with Chrome and Firefox on desktop, and in the android app on a device with an older webview version
2. Record the size of the service worker cache
3. Go offline and check that the app still works, no requests are made on refresh, branding and translations work.
4. Upgrade gamma to the PR branch
5. Go online again and sync and ensure you get prompt to refresh
6. Refresh and check that everything works
7. Record the size of the service worker cache, in particular that the old cache is deleted
8. Go offline and check that the app still works, no requests are made on refresh, branding and translations work.
9. Update branding
10. Go online and sync branding change
11. Go offline and make sure the branding change sticks

**Comment by garethbowen** [comment]
### Test results

The test plan above passed.

Device details:

1. Android OS 6, Webview 87, cht-android 1.1.0
2. Linux, Chromium 116
3. Linux, Firefox 117

Before upgrade:

- CHT 4.1
- 13.6MB Cache storage
- 8.7MB IndexedDB
- 11.47kB Service workers
- 22.3MB total
- Cache Storage name: sw-precache-v3-cache-https://gamma.dev.medicmobile.org

After upgrade:

- CHT 4.3 (branch)
- 14.3MB Cache storage
- 9.5MB IndexedDB
- 17.7kB Service workers
- 23.9MB total
- Cache Storage name: cht-precache-v2-https://gamma.dev.medicmobile.org

Everything worked as expected. Slight increases in storage size are within acceptable range. The old cache storage was deleted on upgrade. The upgrade itself took a few seconds the first time which might be because of deleting the old storage, but then everything was back to normal speed.
